### PR TITLE
Feature cache runs

### PR DESCRIPTION
**Pull request by hardbyte**
This builds upon  https://github.com/data61/anonlink-entity-service/pull/431 to track run state in the redis cache.

This should address #421

### CODE REVIEW

File: backend/entityservice/cache/active_runs.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from entityservice.cache.connection import connect_to_redis
+from entityservice.cache.helpers import _get_run_hash_key
+
+logger = structlog.get_logger()
+
+
+class RunState(Enum):
+    MISSING = None
+    ACTIVE = b'active'
+    DELETED = b'deleted'
```
Comment (gusmith) [inline backend/entityservice/cache/active_runs.py]: Need documentation about these states: what is an active run? Does that mean it is running? What about finished runs?

File: backend/entityservice/cache/active_runs.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+def set_run_state_deleted(run_id):
+    return _set_run_state(run_id, state=RunState.DELETED)
+
+
+def set_run_state_active(run_id):
+    return _set_run_state(run_id, state=RunState.ACTIVE)
+
+
+def is_run_active(run_id):
+    return RunState.ACTIVE == _get_run_state(run_id)
```
Comment (gusmith) [inline backend/entityservice/cache/active_runs.py]: If the run state is missing from the cache, shouldn't we check the information from the database?
Which leads to another question: how long are we keeping this information in redis? Shouldn't we clean the inactive keys? Should we set an expiry time for the keys?

File: backend/entityservice/views/run/description.py [hunk 3]
```diff
@@ -31,6 +32,7 @@ def get(project_id, run_id):
 def _delete_run(run_id, log):
     # Retry if a db integrity error occurs (e.g. when a worker is writing to the same row)
     # Randomly wait up to 2^x * 1 seconds between each retry until the range reaches 60 seconds, then randomly up to 60 seconds afterwards
+    set_run_state_deleted(run_id)
```
Comment (gusmith) [inline backend/entityservice/views/run/description.py]: Are we doing anything different if the run is missing or deleted? If no, could we simply delete the key from redis?

File: backend/entityservice/cache/active_runs.py (hunk 2 above)
Comment (hardbyte) [reply backend/entityservice/cache/active_runs.py]: I'm happy for redis to be the source of truth - after all the entire job is screwed if we lose our celery queue which is stored in Redis. In any case such a check wouldn't be here in the `cache.py` file.

Sure we should we delete keys from redis at some point, not entirely sure it make sense now though. Given we are talking about a handful of bytes per run and by default we provision redis-ha with 1.5GiB of memory... I don't see any great urgency. Would you be happy if I put in an expiry time of a 10 days?

File: backend/entityservice/cache/active_runs.py (hunk 1 above)
Comment (hardbyte) [reply backend/entityservice/cache/active_runs.py]: Sure I'll add a docstring but I'll point out this isn't exposed outside of this file - it is simply to support the public api of: `is_run_active`, `set_run_state_active` and `set_run_state_deleted`.

We could add (and expose) states like "finished" - if and when required.

File: backend/entityservice/views/run/description.py (hunk 3 above)
Comment (hardbyte) [reply backend/entityservice/views/run/description.py]: A missing run means something has gone wrong where a deletion is normal behavior. 

I could be convinced but I'd prefer to set the state in redis so the backend tasks can act accordingly - perhaps throw a big tantrum if the run is missing, but just exit gracefully if the run has been deleted.

File: backend/entityservice/cache/active_runs.py (hunk 2 above)
Comment (hardbyte) [reply backend/entityservice/cache/active_runs.py]: I will however add a `clear_run_state` function to align with the `entityservice.cache.progress` api.

File: backend/entityservice/cache/active_runs.py (hunk 1 above)
Comment (gusmith) [reply backend/entityservice/cache/active_runs.py]: I was also wondering yesterday, should we include an `ERROR` state?

File: backend/entityservice/cache/active_runs.py (hunk 2 above)
Comment (gusmith) [reply backend/entityservice/cache/active_runs.py]: A default of 10 days or more is perfect (configurable even better :) ).
It's mainly to start having best practices when using redis.

**Comment by gusmith** [comment]
Also thinking that we should have at one point some documentation about the different key values stored in redis.

File: backend/entityservice/cache/active_runs.py (hunk 1 above)
Comment (hardbyte) [reply backend/entityservice/cache/active_runs.py]: Perhaps at some point, but I don't want to prematurely add things to the cache that are not used.

File: deployment/entity-service/templates/configmap.yaml [hunk 4]
```diff
@@ -59,9 +59,7 @@ data:
   CHUNK_SIZE_AIM: {{ required "workers.CHUNK_SIZE_AIM is required." .Values.workers.CHUNK_SIZE_AIM | quote }}
 
   MAX_CACHE_SIZE: {{ required "workers.MAX_CACHE_SIZE is required." .Values.workers.MAX_CACHE_SIZE | quote }}
-
-  # ENTITY_MATCH_THRESHOLD not provided
```
Comment (gusmith) [inline deployment/entity-service/templates/configmap.yaml]: Why did you delete this two lines? The two variables are still in use in the file `settings.py`.

File: backend/entityservice/settings.py [hunk 5]
```diff
@@ -52,7 +53,10 @@ class Config(object):
     CELERY_RESULT_BACKEND_TRANSPORT_OPTIONS = {'master_name': "mymaster"} if REDIS_USE_SENTINEL else {}
 
     CELERY_ANNOTATIONS = {
-        'async_worker.calculate_mapping': {'rate_limit': '1/s'}
+        # Note that these are per worker instance per task rate limits.
```
Comment (gusmith) [inline backend/entityservice/settings.py]: Is that why it was taking forever to empty the queue?

File: docs/development.rst [hunk 6]
```diff
@@ -32,6 +32,35 @@ variables. Multiple workers can be used to distribute the work beyond
...
+Redis Cache: Run Info
+~~~~~~~~~~~~~~~~~~~~~
+
+The run info ``HASH`` stores:
+
+- similarity scoring progress for each run under ``"progress"``
+- run state under ``"state"``, current valid states are
+  ``{active, complete, deleted}``. See
+  ``backend/entityservice/cache/active_runs.py`` for implementation.
+
```
Comment (gusmith) [inline docs/development.rst]: :+1:

File: backend/entityservice/tasks/run.py [hunk 7]
```diff
@@ -25,22 +26,26 @@ def prerun_check(project_id, run_id, parent_span=None):
...
             return
 
-        if state in {'running', 'completed', 'error'}:
+        if db_state in {'running', 'completed', 'error'}:
             log.warning("Run already started. Skipping")
             return
 
-        log.debug("Setting run as in progress")
+        # being very defensive here checking if the run state is already in the redis cache
+        if not is_run_missing(run_id):
```
Comment (gusmith) [inline backend/entityservice/tasks/run.py]: Small comment: shouldn't we first check the cache and then the db check? Like that if the check fail with the cache, we don't connect to the db which takes longer. But not important as it is used a single time per run.

File: backend/entityservice/utils.py [hunk 8]
```diff
@@ -124,16 +124,17 @@ def generate_code(length=24):
...
             parties_contributed = get_number_parties_ready(conn, project_id)
-            logger.info("Parties where data is ready: {}".format(parties_contributed))
+            log.info("Parties where data is ready: {}".format(parties_contributed))
         else:
             parties_contributed = get_number_parties_uploaded(conn, project_id)
-            logger.info("Parties where data is uploaded: {}".format(parties_contributed))
+            log.info("Parties where data is uploaded: {}".format(parties_contributed))
         number_parties = get_project_column(conn, project_id, 'parties')
-    logger.info("{}/{} parties have contributed clks".format(parties_contributed, number_parties))
+    log.info("{}/{} parties have contributed clks".format(parties_contributed, number_parties))
```
Comment (gusmith) [inline backend/entityservice/utils.py]: Quick question: should we open an issue to be sure that all methods using logging are using project and run id when available?

File: deployment/entity-service/templates/configmap.yaml (hunk 4 above)
Comment (hardbyte) [reply deployment/entity-service/templates/configmap.yaml]: This setting hasn't had any effect for a long time as the `threshold` is required when posting a new run - see spec [here](https://anonlink-entity-service.readthedocs.io/en/latest/api.html#operation/entityservice.views.run.list.post) - I'll create an issue (https://github.com/data61/anonlink-entity-service/issues/443) and aim to  remove it from settings too.

File: backend/entityservice/settings.py (hunk 5 above)
Comment (hardbyte) [reply backend/entityservice/settings.py]: No I don't think so because the `calculate_mapping` task was the only one rate limited and there actually isn't any task called that any more - hence the small correction to a correct path but commented out.

File: backend/entityservice/tasks/run.py (hunk 7 above)
Comment (hardbyte) [reply backend/entityservice/tasks/run.py]: Fair point, I'll reorder.

I was actually wondering about using a redis lock for this whole section but didn't want to introduce it here.

File: backend/entityservice/utils.py (hunk 8 above)
Comment (hardbyte) [reply backend/entityservice/utils.py]: Sure, I just try improve this kind of thing as I go
